src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java

```diff
@@ -108,4 +150,73 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
     protected String parserName() {
         return IdsQueryParser.NAME;
     }
+
+    public Query toQuery(QueryParseContext parseContext) throws IOException, QueryParsingException {
+        if (this.ids.isEmpty()) {
+            return Queries.newMatchNoDocsQuery();
+        }
+
+        Collection<String> typesForQuery = this.types;
+        if (typesForQuery == null || typesForQuery.isEmpty()) {
+            typesForQuery = parseContext.queryTypes();
+        } else if (typesForQuery.size() == 1 && Iterables.getFirst(typesForQuery, null).equals("_all")) {
+            typesForQuery = parseContext.mapperService().types();
+        }
+
+        List<BytesRef> ids = new ArrayList<>(this.ids.size());
+        for (String value : this.ids) {
+            BytesRef ref = new BytesRef(value);
+            ids.add(ref);
+        }
+
+        TermsQuery query = new TermsQuery(UidFieldMapper.NAME, Uid.createTypeUids(typesForQuery, ids));
+        query.setBoost(boost);
+        if (queryName != null) {
+            parseContext.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        // all fields can be empty or null
+        return null;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        this.types = (List<String>) in.readGenericValue();
+        this.ids = (List<String>) in.readGenericValue();
```

Comment:
I mean both read/writeGenericValue. They are supposed to be used for an object and the method contains a lot of instanceof checks. I would love to call a method that accepts a List, thats it, cause we know upfront the type. The difference between the loops and a specialized method is less lines of code everytime you have a list that you need to serialize, the loop is going to be in the utility method. Makes sense?